apps/librelingo_audios/librelingo_audios/generate_audios.py

```diff
@@ -0,0 +1,88 @@
+from librelingo_utils import audio_id, iterate_phrases
+from pathlib import Path
+import subprocess
+import json
+import random
+
+def generate_audios_for_course(output_path, course_name, course, settings):
+    if not course.settings.audio_settings.enabled:
+        return
+
+    index_path = Path(Path(output_path) / "{}.json".format(course_name))
+
+    phrases_in_course = [(_index_dict_for_phrase(p, course), p) for p in iterate_phrases(course)]
+
+    if settings.overwrite:
+        # Pretend nothing already exists; recreate everything
+        phrase_indices_already_present = []
+    else:
+        with open(index_path, "r") as f:
+            phrase_indices_already_present = json.loads(f.read())["audios"]
+
+    result_index = []
+
+    # For each phrase in the course, if it doesn't exist, generate it.
+    for course_phrase_index, course_phrase in phrases_in_course:
+        # TODO check for equality along multiple dimensions when there are
+        # multiple sources
+        existing_matching_index = next((i for i in phrase_indices_already_present if i["id"] == course_phrase_index["id"]), None)
+
+        if existing_matching_index is None:
+            _generate_audio_for_phrase(course_phrase, output_path, course, settings)
+            result_index.append(course_phrase_index)
+        else:
+            result_index.append(existing_matching_index)
+
+    # For phrases that already exist as audio, check if they are still needed by
```

Comment:
shouldn't this happen before we are generating the audio?

Maybe we could have a test case where in the test setup, we have:

- (A) an audio file that has never been in the meta data and is not needed
- (B) an audio file that's in the meta data now, but the meta data is outdated and after updating it, the file is now not necessary
- (C) an audio file that is still needed and is correctly supplied
- (D) a missing audio file
- (E) an audio file that's needed and exists, but it's missing from the meta file

We can mock the audio file generation so that it always returns with a fake audio file that's based on the original phrase (for example, the content is sth like `f"fake audio file for phrase '{phrase_text}'"`

the assertions should be:

- (A) is deleted
- (B) is also deleted
- (C) remains unchanged
- (D) is generated using the fake audio generation function
- (E) is generated using the fake audio generation function, because do to the lack of data in the metadata, we need to suspect that it's an incorrect file or lacking license

Reply:
I'm working on implementing most of these tests. However, it does not behave exactly how you have outlined here. The index JSON file is the *single source of truth* and the files themselves are a write-only side effect -- they are never read to determine state. So A would not be touched, and D and E are exactly the same from the script's perspective.

Reply:
> So A would not be touched

why, if the JSON file tells you which files should exist, and it tells you nothing about A, it kinda follows that A should not exist, no? 

> D and E are exactly the same from the script's perspective

E should be a similar case as A, basically since it's a file that should not have existed in the first place, it should be removed by the same logic that removed A. Afterwards it should be generated like any other non-existent audio

Reply:
Also, I was thinking that deleting the unneeded files could be a separate cleanup step, but, after some deliberation I think it really should be built into the fetching command.

This is because there are 2 very good reasons to not have any audio files that are not justified by the metadata json:

- The audios might change a lot when changes are merged to a course. This could create a lot of unused files which would be making deployment unnecessarily inefficient
- We want to avoid having unlicensed audio in the course contents hence any files that are not accounted for should be deleted asap

that being said, if you don't feel like implementing the "cleanup" part in this MR, we can create a new issue for it

Reply:
> why, if the JSON file tells you which files should exist, and it tells you nothing about A, it kinda follows that A should not exist, no?

If we've accepted that nobody should be manually editing the `voice` directory, it will be very unusual (and essentially undefined behavior) if anything besides the script has edited it. And no matter how hard we try, it will *never* be possible to ensure correctness despite edits by other parties -- imagine if someone replaced an audio file with a different file but kept the name the same. So, the added "feature" of handling random changes to the file system will be almost never used, and can never work perfectly anyway.

There is also a considerable benefit to the approach I've taken here, in which the directory state is never read. Keeping two things in sync (phrases in the course, phrases in the index) is a surprisingly deep problem, and you'll see in my code that it is already not simple to do so correctly. Adding a third read-write state (audio files in directory) would add a bunch of code complexity, a lot more tests, more surface area for bugs, etc.

> The audios might change a lot when changes are merged to a course. This could create a lot of unused files which would be making deployment unnecessarily inefficient

These would be deleted, though. If the files were created by the script, they will be in the index file. If they are in the index file, they will be deleted if no longer necessary.

> We want to avoid having unlicensed audio in the course contents hence any files that are not accounted for should be deleted asap

Again, this is undefined behavior/user error -- and on top of that, it would not be possible for this audio to ever be actually played in the course, because if it is not in the metadata file that means it is not referenced in the course either (otherwise it would be overwritten by a newly generated file).

Reply:
> E should be a similar case as A, basically since it's a file that should not have existed in the first place, it should be removed by the same logic that removed A. Afterwards it should be generated like any other non-existent audio

For practical purposes, there is no difference between deleting + re-generating and simply overwriting.

Reply:
> For practical purposes, there is no difference between deleting + re-generating and simply overwriting.

that's true

Reply:
> These would be deleted, though. If the files were created by the script, they will be in the index file. If they are in the index file, they will be deleted if no longer necessary.

ok, that makes sense! Looking at the directories then probably doesn't have to be part of this script as your say. _perhaps_ it could be part of a completely separate cleanup script, or not even that.